compiler/noirc_frontend/src/elaborator/traits.rs

```diff
@@ -0,0 +1,183 @@
+        for (trait_id, unresolved_trait) in &traits {
+            self.interner.push_empty_trait(*trait_id, unresolved_trait);
+        }
+
+        for (trait_id, unresolved_trait) in traits {
+            let generics = vecmap(&unresolved_trait.trait_def.generics, |_| {
+                TypeVariable::unbound(self.interner.next_type_variable_id())
+            });
+
+            // Resolve order
+            // 1. Trait Types ( Trait constants can have a trait type, therefore types before constants)
+            let _ = self.resolve_trait_types(&unresolved_trait);
+            // 2. Trait Constants ( Trait's methods can use trait types & constants, therefore they should be after)
+            let _ = self.resolve_trait_constants(&unresolved_trait);
+            // 3. Trait Methods
+            let methods =
+                self.resolve_trait_methods(trait_id, &unresolved_trait, &generics);
+
+            self.interner.update_trait(trait_id, |trait_def| {
+                trait_def.set_methods(methods);
+                trait_def.generics = generics;
+            });
+
+            // This check needs to be after the trait's methods are set since
+            // the interner may set `interner.ordering_type` based on the result type
+            // of the Cmp trait, if this is it.
+            if self.crate_id.is_stdlib() {
+                self.interner.try_add_operator_trait(trait_id);
+            }
+        }
+    }
+
+    fn resolve_trait_types(
+        &mut self,
+        _unresolved_trait: &UnresolvedTrait,
+    ) -> Vec<TraitType> {
+        // TODO
+        vec![]
+    }
+
+    fn resolve_trait_constants(
+        &mut self,
+        _unresolved_trait: &UnresolvedTrait,
+    ) -> Vec<TraitConstant> {
+        // TODO
+        vec![]
+    }
+
+    fn resolve_trait_methods(
+        &mut self,
+        trait_id: TraitId,
+        unresolved_trait: &UnresolvedTrait,
+        trait_generics: &Generics,
+    ) -> Vec<TraitFunction> {
+        self.local_module = unresolved_trait.module_id;
+        self.file = self.def_maps[&self.crate_id].file_id(unresolved_trait.module_id);
+
+        let mut functions = vec![];
+
+        for item in &unresolved_trait.trait_def.items {
+            if let TraitItem::Function {
+                name,
+                generics,
+                parameters,
+                return_type,
+                where_clause,
+                body: _,
+            } = item
+            {
+                let old_generic_count = self.generics.len();
+
+                let the_trait = self.interner.get_trait(trait_id);
+                let self_typevar = the_trait.self_type_typevar.clone();
+                let self_type = Type::TypeVariable(self_typevar.clone(), TypeVariableKind::Normal);
+                let name_span = the_trait.name.span();
+
+                self.add_generics(generics);
+                self.add_existing_generics(&unresolved_trait.trait_def.generics, trait_generics);
```

Comment:
I think `existing` here refers to 'instantiated' or 'resolved' generics?

Reply:
`existing` refers to resolved generics yeah. Generics that have previously already been added to the name resolver or elaborator in this case but were removed at some point and need to be re-added.